File: golang/advent-of-code-2021/two/two.go
package two

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Command struct {
	Direction string
	Step      int
}

func ReadCommandFromFile() []Command {
	file, err := os.Open("two/two.txt")
	if err != nil {
		panic("Cannot find/open input file for day 2")
	}

	reader := bufio.NewReader(file)
	commands := make([]Command, 0)
	for {
		currentLine, _, err := reader.ReadLine()
		if err != nil {
			break
		}

		parts := strings.Split(string(currentLine), " ")
		step, err := strconv.Atoi(parts[1])
		if err != nil {
			fmt.Printf("Cannot convert %s to number\n", parts[1])
		}

		commands = append(commands, Command{Direction: parts[0], Step: step})
	}

	return commands
}

func SimulateCommands(commands *[]Command) (int, int) {
	horizontal := 0
	vertical := 0
	for _, command := range *commands {
		switch command.Direction {
		case "forward":
			horizontal += command.Step
		case "up":
			vertical -= command.Step
		case "down":
			vertical += command.Step
		}
	}

	return horizontal, vertical
}

func SimulateCommandsWithAim(commands *[]Command) (int, int) {
	horizontal, vertical, aim := 0, 0, 0
	for _, command := range *commands {
		switch command.Direction {
		case "forward":
			horizontal += command.Step
			vertical += aim * command.Step
		case "up":
			aim -= command.Step
		case "down":
			aim += command.Step
		}
	}

	return horizontal, vertical
}

File: rust/advent-of-code-2021/src/solutions/two.rs
use std::{
    fs::File,
    io::{BufRead, BufReader},
};

pub enum Direction {
    Forward,
    Down,
    Up,
}

pub struct Command {
    direction: Direction,
    steps: usize,
}

pub fn read_commands_from_file() -> Vec<Command> {
    let file = File::open("src/inputs/two.txt").expect("Cannot find/open input file for day2");
    let reader = BufReader::new(file);

    let mut commands: Vec<Command> = vec![];
    for line in reader.lines() {
        let current_line = line.expect("Cannot read current line. Aborting...");

        let parts: Vec<&str> = current_line.split(" ").collect();
        let direction = match parts[0] {
            "forward" => Direction::Forward,
            "down" => Direction::Down,
            "up" => Direction::Up,
            _ => Direction::Forward,
        };

        commands.push(Command {
            direction,
            steps: parts[1]
                .parse()
                .expect("Cannot parse step from string to usize"),
        });
    }

    return commands;
}

pub fn simulate_commands(commands: &Vec<Command>) -> (isize, isize) {
    let mut x: isize = 0;
    let mut y: isize = 0;
    for command in commands {
        let value: isize = isize::try_from(command.steps).unwrap();
        match command.direction {
            Direction::Forward => x += value,
            Direction::Down => y += value,
            Direction::Up => y -= value,
        };
    }

    (x, y)
}

pub fn simulate_commands_with_aim(commands: &Vec<Command>) -> (isize, isize) {
    let mut x: isize = 0;
    let mut y: isize = 0;
    let mut aim: isize = 0;
    for command in commands {
        let value: isize = isize::try_from(command.steps).unwrap();
        match command.direction {
            Direction::Forward => {
                x += value;
                y += value * aim;
            }
            Direction::Down => aim += value,
            Direction::Up => aim -= value,
        };
    }

    (x, y)
}
